platforms/nuttx/src/px4/rpi/rpi_common/hrt/hrt.c

```diff
@@ -0,0 +1,809 @@
+		ppm.last_mark = ppm.last_edge;
+
+		/* frame length is everything including the start gap */
+		ppm_frame_length = (uint16_t)(ppm.last_edge - ppm.frame_start);
+		ppm.frame_start = ppm.last_edge;
+		ppm.phase = ACTIVE;
+		break;
+
+	case INACTIVE:
+
+		/* we expect a short pulse */
+		if (width < PPM_MIN_PULSE_WIDTH || width > PPM_MAX_PULSE_WIDTH) {
+			goto error;        /* pulse was too short or too long */
+		}
+
+		/* this edge is not interesting, but now we are ready for the next mark */
+		ppm.phase = ACTIVE;
+		break;
+
+	case ACTIVE:
+		/* determine the interval from the last mark */
+		interval = count - ppm.last_mark;
+		ppm.last_mark = count;
+
+#if PPM_DEBUG
+		ppm_pulse_history[ppm_pulse_next++] = interval;
+
+		if (ppm_pulse_next >= 32) {
+			ppm_pulse_next = 0;
+		}
+
+#endif
+
+		/* if the mark-mark timing is out of bounds, abandon the frame */
+		if ((interval < PPM_MIN_CHANNEL_VALUE) || (interval > PPM_MAX_CHANNEL_VALUE)) {
+			goto error;
+		}
+
+		/* if we have room to store the value, do so */
+		if (ppm.next_channel < PPM_MAX_CHANNELS) {
+			ppm_temp_buffer[ppm.next_channel++] = interval;
+		}
+
+		ppm.phase = INACTIVE;
+		break;
+
+	}
+
+	ppm.last_edge = count;
+	return;
+
+	/* the state machine is corrupted; reset it */
+
+error:
+	/* we don't like the state of the decoder, reset it and try again */
+	ppm.phase = UNSYNCH;
+	ppm_decoded_channels = 0;
+
+}
+#endif /* HRT_PPM_CHANNEL */
+
+/**
+ * Handle the compare interrupt by calling the callout dispatcher
+ * and then re-scheduling the next deadline.
+ */
+static int
+hrt_tim_isr(int irq, void *context, void *arg)
+{
+	/* grab the timer for latency tracking purposes */
+	latency_actual = hrt_absolute_time();
+
+	rINTR = rINTR;	// ack the interrupts we just read
+
+	/* It is never a timer tick. It is always triggered by alarm. */
+	/* do latency calculations */
+	hrt_latency_update();
+
+	/* run any callouts that have met their deadline */
+	hrt_call_invoke();
+
+	/* and schedule the next interrupt */
+	hrt_call_reschedule();
+
+	return OK;
+}
+
+/**
+ * Handle the compare interrupt by calling the callout dispatcher
+ * and then re-scheduling the next deadline.
+ */
+static int
+hrt_ppm_isr(int irq, void *context, void *arg)
+{
+	// Read lower 16 bits of hrt
+	uint16_t counter = rTIMERAWL & 0xffff;
+
+	// Switch the next interrupt type
+	px4_arch_gpiosetevent(GPIO_PPM_IN, lastEdge ? false : true, lastEdge ? true : false, true, hrt_ppm_isr, NULL);
+	lastEdge = !lastEdge;
+
+	hrt_ppm_decode(counter);
+	return OK;
+}
+
+/**
+ * Fetch a never-wrapping absolute time value in microseconds from
+ * some arbitrary epoch shortly after system start.
+ */
+hrt_abstime hrt_absolute_time(void)
+{
+	/* Taken from rp2040 datasheet pg. 558 */
+	uint32_t hi = rTIMERAWH;
+	uint32_t lo;
+
+	do {
+		lo = rTIMERAWL;
+		uint32_t next_hi = rTIMERAWH;
+
+		if (hi == next_hi) { break; }
+
+		hi = next_hi;
+	} while (true);
+
+	return ((uint64_t) hi << 32) | lo;
+}
```

Comment:
So cool that you literally had to implement the `hrt_absoluate_time` function haha @vxj9800 

I was going through the PR as I was watching your talk on Youtube, super cool!

Reply:
@junwoo091400 This was a great relief when I started working on this since the RP2040 has a dedicated 64-bit real-time timer. It made my life much easier.